File: PixelSolution/PixelEngine/Group.cpp
#include "pch.h"
#include "Group.h"
#include "PixelEngine.h"
#include "SceneObject.h"
#include "SceneManager.h"
#include "GroupManager.h"

extern PixelEngine* Engine;
unsigned int Group_CreateGroupAndEntity(const char* group, const char* scriptName)
{
	SceneManager* scene = Engine->GetFactory<SceneManager>();
	std::string groupNameSTR(group);
	std::string scriptNameSTR(scriptName);
	return scene->GetNowScene()->CreateGroupAndEntity(groupNameSTR, scriptNameSTR);
}

void Group_Set(const char* groupName, unsigned int id)
{
	SceneManager* scene = Engine->GetFactory<SceneManager>();
	std::string groupNameSTR(groupName);
	scene->GetNowScene()->GetGroupManager()->Set(groupNameSTR, id);
}

void Group_Remove(const char* groupName, unsigned int id)
{
	SceneManager* scene = Engine->GetFactory<SceneManager>();
	std::string groupNameSTR(groupName);
	scene->GetNowScene()->GetGroupManager()->Remove(groupNameSTR, id);
}

void Group_Clear(const char* groupName)
{
	SceneManager* scene = Engine->GetFactory<SceneManager>();
	std::string groupNameSTR(groupName);
	scene->GetNowScene()->GetGroupManager()->Clear(groupNameSTR);
}

int Group_Count(const char* groupName)
{
	SceneManager* scene = Engine->GetFactory<SceneManager>();
	std::string groupNameSTR(groupName);
	auto K = scene->GetNowScene()->GetGroupManager()->Get(groupNameSTR);
	return (int)K->size();
}

unsigned int Group_First(const char* groupName)
{
	SceneManager* scene = Engine->GetFactory<SceneManager>();
	std::string groupNameSTR(groupName);
	std::vector<unsigned int>* K = scene->GetNowScene()->GetGroupManager()->Get(groupNameSTR);
	if (K->size() != 0)
	{
		return (*K)[0];
	}
}

int Group_GetCount(const char* groupName)
{
	SceneManager* scene = Engine->GetFactory<SceneManager>();
	std::string groupNameSTR(groupName);
	auto ids = scene->GetNowScene()->GetGroupManager()->Get(groupNameSTR);
	if (ids == nullptr)
	{
		return 0;
	}

	return static_cast<int>(ids->size());
}

unsigned int Group_GetID(const char* groupName, int index)
{
	SceneManager* scene = Engine->GetFactory<SceneManager>();
	std::string groupNameSTR(groupName);
	auto ids = scene->GetNowScene()->GetGroupManager()->Get(groupNameSTR);
	if (ids == nullptr || index < 0 || index >= static_cast<int>(ids->size()))
	{
		return 0;
	}

	return (*ids)[index];
}

File: Survival/Engine/Group.lua
local ffi = require("ffi")
local dll = ffi.load("PixelEngine")

---@class Group
ffi.cdef[[
	unsigned int Group_CreateGroupAndEntity(const char* GroupName,const char* ScriptName);
	void Group_Set(const char* GroupName,unsigned int ID);
	void Group_Remove(const char* GroupName,unsigned int ID);
	void Group_Clear(const char* GroupName);
	int Group_Count(const char* GroupName);
	unsigned int Group_First(const char* GroupName);
	int Group_GetCount(const char* GroupName);
	unsigned int Group_GetID(const char* GroupName,int Index);
]]

---@class Group
Group = Group or {}

---@param GroupName string
---@param ScriptName string
---@return number
function Group.CreateGroupAndEntity(GroupName,ScriptName)
	return dll.Group_CreateGroupAndEntity(GroupName,ScriptName)
end
---@param GroupName string
---@param ID number
function Group.Set(GroupName,ID)
	return dll.Group_Set(GroupName,ID)
end
---@param GroupName string
---@param ID number
function Group.Remove(GroupName,ID)
	return dll.Group_Remove(GroupName,ID)
end
---@param GroupName string
function Group.Clear(GroupName)
	return dll.Group_Clear(GroupName)
end
---@param GroupName string
---@return number
function Group.Count(GroupName)
	return dll.Group_Count(GroupName)
end
---@param GroupName string
---@return number
function Group.First(GroupName)
	return dll.Group_First(GroupName)
end
---@param GroupName string
---@return number
function Group.GetCount(GroupName)
	return dll.Group_GetCount(GroupName)
end
---@param GroupName string
---@param Index number
---@return number
function Group.GetID(GroupName,Index)
	return dll.Group_GetID(GroupName,Index)
end
return Group